tty.c

```diff
@@ -214,48 +348,82 @@ void tty_parent_setup(int socket) {
 	if (info.stdinIsatty) {
 		info.orig = tios;
 		cfmakeraw(&tios);
-		if (tcsetattr(STDIN_FILENO, TCSANOW, &tios) < 0) {
+		if (tcsetattr(STDIN_FILENO, TCSANOW, &tios) == -1) {
 			err(1, "tty_parent: tcsetattr");
 		}
 	}
-	atexit(tty_parent_cleanup);
 
 	// Wait for the child to create the pty pair and pass the master back.
 	recv_fd(socket, &info.termfd);
 
 	if (info.stdinIsatty) {
-		if (tcsetattr(info.termfd, TCSAFLUSH, &info.orig) < 0) {
+		if (tcsetattr(info.termfd, TCSAFLUSH, &info.orig) == -1) {
 			err(1, "tty_parent: tcsetattr");
 		}
 	}
 
 	sigset_t sigmask;
-	sigfillset(&sigmask);
-	if (sigprocmask(SIG_BLOCK, &sigmask, NULL) < 0) {
-		err(1, "tty_parent: sigprocmask");
-	}
-	if ((info.sigfd = signalfd(-1, &sigmask, 0)) < 0) {
+	sigemptyset(&sigmask);
+	sigaddset(&sigmask, SIGWINCH);
+
+	int sigfd = signalfd(-1, &sigmask, 0);
+	if (sigfd == -1) {
 		err(1, "tty_parent: signalfd");
 	}
-	FD_ZERO(&info.rfds);
-	FD_ZERO(&info.wfds);
-	FD_SET(STDIN_FILENO, &info.rfds);
-	FD_SET((unsigned long) info.termfd, &info.rfds);
-	FD_SET((unsigned long) info.sigfd, &info.rfds);
-	FD_SET(STDOUT_FILENO, &info.wfds);
-	FD_SET((unsigned long) info.termfd, &info.wfds);
-	if (info.sigfd > info.termfd) {
-		info.nfds = info.sigfd + 1;
-	} else {
-		info.nfds = info.termfd + 1;
+
+	static struct epoll_handler sighandler;
+	sighandler.fn = tty_handle_sig;
+	sighandler.fd = sigfd;
+
+	struct epoll_event event = {
+		.events = EPOLLIN,
+		.data.ptr = &sighandler,
+	};
+
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, sigfd, &event) == -1) {
+		err(1, "epoll_ctl_add signalfd");
+	}
+
+	inbound_handler.fn = tty_handle_io;
```

Comment:
NIT: set fd & fd with a designated initializer on inbound/outbound handler so we only have to worry about peer_fd here.